DataStructures/BinaryHeap.cs

```diff
@@ -0,0 +1,251 @@
+    public class BinaryHeap<T>
+    {
+        /// <summary>
+        /// Comparer to use when comparing elements.
+        /// </summary>
+        private readonly Comparer<T> comparer;
+
+        /// <summary>
+        /// List to hold the elements of the heap.
+        /// </summary>
+        private readonly List<T> data;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BinaryHeap{T}"/> class.
+        /// </summary>
+        public BinaryHeap()
+        {
+            data = new List<T>();
+            comparer = Comparer<T>.Default;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BinaryHeap{T}"/> class with a custom comparision function.
+        /// </summary>
+        /// <param name="customComparer">The custom comparing function to use to compare elements.</param>
+        public BinaryHeap(Comparer<T> customComparer)
+        {
+            data = new List<T>();
+            comparer = customComparer;
+        }
+
+        /// <summary>
+        /// Gets the number of elements in the heap.
+        /// </summary>
+        public int Count => data.Count;
+
+        /// <summary>
+        /// Add an element to the binary heap.
+        /// </summary>
+        /// <remarks>
+        /// Adding to the heap is done by append the element to the end of the backing list,
+        /// and pushing the added element up until the heap property is restored.
+        /// </remarks>
+        /// <param name="element">The element to add to the heap.</param>
+        /// <exception cref="ArgumentException">Thrown if element is already in heap.</exception>
+        public void Push(T element)
+        {
+            if (data.Contains(element))
+            {
+                throw new ArgumentException($"{element} already in heap!");
+            }
+
+            data.Add(element);
+            HeapifyUp(data.Count - 1);
+        }
+
+        /// <summary>
+        /// Remove the top/root of the binary heap (ie: the largest/smallest element).
+        /// </summary>
+        /// <returns>The top/root of the heap.</returns>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        public T Pop()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty!");
+            }
+
+            T elem = data[0];
+            data[0] = data[data.Count - 1];
+            data.RemoveAt(data.Count - 1);
+            HeapifyDown(0);
+
+            return elem;
+        }
+
+        /// <summary>
+        /// Return the top/root of the heap without remove it.
+        /// </summary>
+        /// <remarks>
+        /// Removing from the heap is done by swapping the top/root with the last element in
+        /// the backing list, removing the last element, and pushing the new root down
+        /// until the heap property is restored.
+        /// </remarks>
```

Comment:
Is this relevant?

Reply:
That section was in the wrong place; I have moved it above the correct function.